File: openclaw-ios/OpenClaw/ViewModels/ChatViewModel.swift
import Foundation
import SwiftUI

@MainActor
class ChatViewModel: ObservableObject {

    @Published var messages: [ChatMessage] = []
    @Published var isLoading = false
    @Published var inputText = ""

    let voiceService = VoiceService()

    init() {
        // When STT delivers result -> send automatically
        voiceService.onResult = { [weak self] text in
            Task { @MainActor in
                self?.sendMessage(text)
            }
        }
    }

    func sendMessage(_ text: String) {
        guard !text.isEmpty else { return }

        // Add user message
        let userMsg = ChatMessage(role: "user", content: text)
        messages.append(userMsg)
        isLoading = true
        inputText = ""

        Task {
            do {
                let response = try await APIService.shared.sendMessage(text)

                let assistantMsg = ChatMessage(
                    id: response.id,
                    role: "assistant",
                    content: response.content,
                    model: response.model
                )
                messages.append(assistantMsg)

                // Speak answer
                voiceService.speak(response.content)

            } catch {
                let errMsg = ChatMessage(
                    role: "assistant",
                    content: "Fehler: \(error.localizedDescription)"
                )
                messages.append(errMsg)
            }
            isLoading = false
        }
    }

    func startListening() {
        Task {
            let granted = await voiceService.requestPermissions()
            if granted {
                voiceService.startListening()
            }
        }
    }

    func stopListening() {
        voiceService.stopListening()
    }
}

File: openclaw-android/app/src/main/java/de/openclaw/assistant/viewmodel/ChatViewModel.kt
package de.openclaw.assistant.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import de.openclaw.assistant.data.api.ApiClient
import de.openclaw.assistant.data.model.*
import de.openclaw.assistant.voice.SpeechToTextManager
import de.openclaw.assistant.voice.TextToSpeechManager
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch

class ChatViewModel(app: Application) : AndroidViewModel(app) {

    val stt = SpeechToTextManager(app)
    val tts = TextToSpeechManager(app)

    private val _messages = MutableStateFlow<List<ChatMessage>>(emptyList())
    val messages: StateFlow<List<ChatMessage>> = _messages

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading

    fun startListening() {
        stt.startListening("de-DE")
    }

    fun stopListening() {
        stt.stopListening()
    }

    init {
        // Wenn STT ein Ergebnis liefert → automatisch senden
        viewModelScope.launch {
            stt.state.collect { state ->
                if (state is SpeechToTextManager.STTState.Result) {
                    sendMessage(state.text)
                }
            }
        }
    }

    fun sendMessage(text: String) {
        if (text.isBlank()) return

        // User-Nachricht hinzufügen
        val userMsg = ChatMessage(
            id = "user_${System.currentTimeMillis()}",
            role = "user",
            content = text
        )
        _messages.value = _messages.value + userMsg
        _isLoading.value = true

        viewModelScope.launch {
            try {
                val response = ApiClient.api.sendMessage(
                    ChatRequest(message = text)
                )
                if (response.isSuccessful) {
                    val body = response.body()!!
                    val assistantMsg = ChatMessage(
                        id = body.id,
                        role = "assistant",
                        content = body.content,
                        model = body.model
                    )
                    _messages.value = _messages.value + assistantMsg

                    // Antwort vorlesen
                    tts.speak(body.content)
                } else {
                    handleError("API Error: ${response.code()}")
                }
            } catch (e: Exception) {
                handleError("Network Error: ${e.message}")
            } finally {
                _isLoading.value = false
            }
        }
    }

    private fun handleError(msg: String) {
        val errMsg = ChatMessage(
            id = "err_${System.currentTimeMillis()}",
            role = "assistant",
            content = "Fehler: $msg"
        )
        _messages.value = _messages.value + errMsg
    }

    override fun onCleared() {
        stt.destroy()
        tts.destroy()
    }
}
